atom/browser/api/atom_api_url_request.h

```diff
@@ -0,0 +1,242 @@
+// strong/weak pointers. A URLRequest instance is deleted if it is unpinned and
+// the corresponding JS wrapper object is garbage collected. On the other hand,
+// an AtmURLRequest instance lifetime is totally governed by reference counting.
+//
+class URLRequest : public mate::EventEmitter<URLRequest> {
+ public:
+  static mate::WrappableBase* New(mate::Arguments* args);
+
+  static void BuildPrototype(v8::Isolate* isolate,
+                             v8::Local<v8::FunctionTemplate> prototype);
+
+  // Methods for reporting events into JavaScript.
+  void OnAuthenticationRequired(
+      scoped_refptr<const net::AuthChallengeInfo> auth_info);
+  void OnResponseStarted(
+      scoped_refptr<net::HttpResponseHeaders> response_headers);
+  void OnResponseData(scoped_refptr<const net::IOBufferWithSize> data);
+  void OnResponseCompleted();
+  void OnError(const std::string& error, bool isRequestError);
+
+ protected:
+  explicit URLRequest(v8::Isolate* isolate, v8::Local<v8::Object> wrapper);
+  ~URLRequest() override;
+
+ private:
+  template <typename Flags>
+  class StateBase {
+   public:
+    void SetFlag(Flags flag);
+
+   protected:
+    explicit StateBase(Flags initialState);
+    bool operator==(Flags flag) const;
+    bool IsFlagSet(Flags flag) const;
+
+   private:
+    Flags state_;
+  };
+
+  enum class RequestStateFlags {
+    kNotStarted = 0x0,
+    kStarted = 0x1,
+    kFinished = 0x2,
+    kCanceled = 0x4,
+    kFailed = 0x8,
+    kClosed = 0x10
+  };
+
+  class RequestState : public StateBase<RequestStateFlags> {
+   public:
+    RequestState();
+    bool NotStarted() const;
+    bool Started() const;
+    bool Finished() const;
+    bool Canceled() const;
+    bool Failed() const;
+    bool Closed() const;
+  };
+
+  enum class ResponseStateFlags {
+    kNotStarted = 0x0,
+    kStarted = 0x1,
+    kEnded = 0x2,
+    kFailed = 0x4
+  };
+
+  class ResponseState : public StateBase<ResponseStateFlags> {
+   public:
+    ResponseState();
+    bool NotStarted() const;
+    bool Started() const;
+    bool Ended() const;
+    bool Canceled() const;
+    bool Failed() const;
+    bool Closed() const;
+  };
+
+  bool NotStarted() const;
+  bool Finished() const;
+  bool Canceled() const;
+  bool Failed() const;
+  bool Write(scoped_refptr<const net::IOBufferWithSize> buffer, bool is_last);
+  void Cancel();
+  bool SetExtraHeader(const std::string& name, const std::string& value);
+  void RemoveExtraHeader(const std::string& name);
+  void SetChunkedUpload(bool is_chunked_upload);
+
+  int StatusCode() const;
+  std::string StatusMessage() const;
+  scoped_refptr<net::HttpResponseHeaders> RawResponseHeaders() const;
+  uint32_t ResponseHttpVersionMajor() const;
+  uint32_t ResponseHttpVersionMinor() const;
+
+  template <typename... ArgTypes>
+  std::array<v8::Local<v8::Value>, sizeof...(ArgTypes)> BuildArgsArray(
+      ArgTypes... args) const;
+
+  template <typename... ArgTypes>
+  void EmitRequestEvent(ArgTypes... args);
```

Comment:
How about moving this to the `.cc` file as a helper function?

And to make the code simpler (and also handle the micro tasks correctly), we can also refactor `event_emitter_caller.h` to make `CallEmitWithArgs` accept  a custom method:

```c++
template<typename... ArgTypes>
void CustomEmit(v8::Isolate* isolate,
                v8::Local<v8::Object> object,
                const ArgTypes&... args) {
  std::vector<v8::Local<v8::Value>> converted_args = {
    mate::ConvertToV8(isolate, args)...
  };
  mate::internal::CallEmitWithArgs(isolate, object, &converted_args);
}
```